include/ddc/kernels/fft.hpp

```diff
@@ -407,203 +201,79 @@ void impl(
-                1,
-                idist,
-                reinterpret_cast<typename _fftw_type<Tout>::type*>(out_data),
-                static_cast<int*>(nullptr),
-                1,
-                odist);
-        if constexpr (std::is_same_v<real_type_t<Tin>, float>) {
-            fftwf_execute(plan);
-            fftwf_destroy_plan(plan);
-        } else {
-            fftw_execute(plan);
-            fftw_destroy_plan(plan);
-        }
-    }
-#endif
-#if fftw_omp_AVAIL
-    if constexpr (std::is_same_v<ExecSpace, Kokkos::OpenMP>) {
-        if constexpr (std::is_same_v<real_type_t<Tin>, float>) {
-            fftwf_init_threads();
-            fftwf_plan_with_nthreads(exec_space.concurrency());
+    Kokkos::View<
+            ddc::detail::mdspan_to_kokkos_element_t<Tin, sizeof...(DDimIn)>,
+            ddc::detail::mdspan_to_kokkos_layout_t<LayoutIn>,
+            MemorySpace> const in_view
+            = in.allocation_kokkos_view();
+    Kokkos::View<
+            ddc::detail::mdspan_to_kokkos_element_t<Tout, sizeof...(DDimIn)>,
+            ddc::detail::mdspan_to_kokkos_layout_t<LayoutOut>,
+            MemorySpace> const out_view
+            = out.allocation_kokkos_view();
+    KokkosFFT::Normalization const kokkos_fft_normalization
+            = ddc_fft_normalization_to_kokkos_fft(kwargs.normalization);
+
+    // C2C
+    if constexpr (std::is_same_v<Tin, Tout>) {
+        if (kwargs.direction == ddc::FFT_Direction::FORWARD) {
+            KokkosFFT::
+                    fftn(exec_space,
+                         in_view,
+                         out_view,
+                         axes<DDimIn...>(),
+                         kokkos_fft_normalization);
         } else {
-            fftw_init_threads();
-            fftw_plan_with_nthreads(exec_space.concurrency());
+            KokkosFFT::
+                    ifftn(exec_space,
+                          in_view,
+                          out_view,
+                          axes<DDimIn...>(),
+                          kokkos_fft_normalization);
         }
-        _fftw_plan<Tin> plan = _fftw_plan_many_dft<Tin, Tout>(
-                kwargs.direction == ddc::FFT_Direction::FORWARD ? FFTW_FORWARD : FFTW_BACKWARD,
-                FFTW_ESTIMATE,
-                static_cast<int>(sizeof...(DDimX)),
-                n.data(),
-                1,
-                reinterpret_cast<typename _fftw_type<Tin>::type*>(in_data),
-                static_cast<int*>(nullptr),
-                1,
-                idist,
-                reinterpret_cast<typename _fftw_type<Tout>::type*>(out_data),
-                static_cast<int*>(nullptr),
-                1,
-                odist);
-        if constexpr (std::is_same_v<real_type_t<Tin>, float>) {
-            fftwf_execute(plan);
-            fftwf_destroy_plan(plan);
+        // R2C & C2R
+    } else {
+        if constexpr (is_complex_v<Tout>) {
+            assert(kwargs.direction == ddc::FFT_Direction::FORWARD);
```

Comment:
if this is an error that might actually happen if the user doesn't check its input and not just in case of a problem in the implementation, wouldn't an Exception be better? A message would be the minimum IMHO

Reply:
I just had a look, actually it is more an artifact of our internal `impl` function that requires a direction, the user has never to pass this variable. We could even remove the assert I guess.